### PR TITLE
Implemented SA1121

### LINKED ISSUE #54: SA1121: UseBuiltInTypeAlias
http://www.stylecop.com/docs/SA1121.html

### PR DESCRIPTION
**Pull request by pdelvo**
Fixes #54. Unfortunatly the diagnostic itself was already implemented in #306. But this one contains a bunch of unit tests so it might be a good idea to use them

### CODE REVIEW

**Comment by sharwell** [comment]
This is currently blocked on #306. See #54 for additional information.

**Comment by sharwell** [comment]
Now that #306 is merged, can you rebase this and reduce it to a nice set of tests (plus enabling the diagnostic)? :+1:

**Comment by pdelvo** [comment]
This should bring in the tests. Im working on the code fix

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1121UnitTests.cs [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+
+    /// <summary>
+    /// This class contains unit tests for <see cref="SA1121UseBuiltInTypeAlias"/>
+    /// </summary>
+    [TestClass]
+    public class SA1121UnitTests : CodeFixVerifier
+    {
+        static readonly Tuple<string, string>[] _referenceTypes = new Tuple<string, string>[]
+        {
+            new Tuple<string,string>("object", "Object"),
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1121UnitTests.cs]: `nameof(Object)` for the right hand side here. :wink: The same for the items below.

**Comment by sharwell** [comment]
Can you add this test?

```csharp
namespace Foo
{
  using Int32 = System.UInt32;
  class Bar
  {
    Int32 value = 3;
  }
}
```

A diagnostic should be reported, but the target built-in type is actually `uint` due to the misleading using alias directive.

**Comment by sharwell** [comment]
Also can you add this test?

```csharp
namespace Foo
{
  using MyInt = System.Int32;
  class Bar
  {
    MyInt value = 3;
  }
}
```

**Comment by pdelvo** [comment]
What are the desired results here? Should using Int32 = System.UInt32; change to using Int32 = uint; or should the Int32 value = 3; change to uint value = 3; ? Or both? Currently it will report 2 diagnostics at MyInt value = 3; and the code fix im implementing is replacing it there

**Comment by sharwell** [comment]
1. Using alias directives are not allowed to use the language-specific keywords to refer to types. No diagnostic should be reported for the following line:

    ```csharp
    using Int32 = System.Int32;
    ```

2. In the first example I posted (the misleading using alias directive), the using alias directive would not be altered by the code fix. The reference to the alias would be changed from `Int32` to `uint`.

3. In the second example I posted, the using alias directive again would not change, and the reference to `MyInt` would be changed to `int`.

**Comment by pdelvo** [comment]
These new tests have some issues. The diagnostic analyzer is called twice for the same IdentifierToken. Do you know why this might happen?

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1121UnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,1114 @@
...
+                        Id = DiagnosticId,
+                        Message = "Use built-in type alias",
+                        Severity = DiagnosticSeverity.Warning,
+                        Locations =
+                            new[]
+                            {
+                                new DiagnosticResultLocation("Test0.cs", 6, 5)
+                            }
+                    },
+                // Workaround because the diagnostic is called twice for the SyntaxNode
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1121UnitTests.cs]: I could not figure out why this happens. Maybe a roslyn bug?

**Comment by sharwell** [comment]
It's a bug somewhere in the unit test framework for this extension (modified from the one in the Roslyn templates). The diagnostics are only reported once for actual code, so I merged this and created an issue to find the bug in the unit test framework.

**Pull request merged**